hstream/src/HStream/Server/Core/Subscription.hs

```diff
@@ -579,24 +578,31 @@ sendRecords ServerContext{..} subState subCtx@SubscribeContext {..} = do
       -- Note: non-strict behaviour in STM!
       --       Please refer to https://github.com/haskell/stm/issues/30
       newTime <- getCurrentTimestamp <&> fromIntegral
-      doneList <- atomically $ do
+      timeoutList <- atomically $ do
         ct <- readTVar subCurrentTime
         checkList <- readTVar subWaitingCheckedRecordIds
-        let (!doneList, !leftList) = L.partition (\CheckedRecordIds {..} -> crDeadline <= newTime) checkList
+        let (!timeoutList, !leftList) = Heap.span (\CheckedRecordIds {..} -> crDeadline <= newTime) checkList
         writeTVar subCurrentTime newTime
-        return doneList
-      forM_ doneList (\r@CheckedRecordIds {..} -> buildShardRecordIds r >>= resendTimeoutRecords crLogId crBatchId )
-      recordIds <- mapM (\CheckedRecordIds {..} -> do
-                           idxs <- readTVarIO crBatchIndexes
-                           return $ L.map (\idx ->
-                                             RecordId {
-                                               recordIdShardId = crLogId
-                                             , recordIdBatchId = crBatchId
-                                             , recordIdBatchIndex = idx
-                                             }
-                                          ) (Set.toList idxs)
-                        ) doneList <&> (V.fromList . L.concat)
+        writeTVar subWaitingCheckedRecordIds leftList
+        return timeoutList
+      recordIds <- foldM
+                      (
+                        \acc CheckedRecordIds {..} -> do
+                          idxs <- readTVarIO crBatchIndexes
+                          let ids =
+                                    V.fromList $  fmap (\idx ->
+                                            RecordId {
+                                              recordIdShardId = crLogId
+                                            , recordIdBatchId = crBatchId
+                                            , recordIdBatchIndex = idx
+                                            }
+                                         ) (Set.toList idxs)
+                          pure $ acc V.++ ids
+                      )
+                      V.empty
+                      timeoutList
       atomically $ removeAckedRecordIdsFromCheckList subCtx recordIds
+      forM_ timeoutList (\r@CheckedRecordIds {..} -> buildShardRecordIds r >>= resendTimeoutRecords crLogId crBatchId )
```

Comment:
I think the removal process should be placed after resending records... Otherwise, it will get stuck.